asmcomp/linscan.ml

```diff
@@ -73,10 +75,27 @@ let rec release_expired_inactive ci pos = function
 
 let allocate_stack_slot num_stack_slots i =
   let cl = Proc.register_class i.reg in
-  let ss = num_stack_slots.(cl) in
-  num_stack_slots.(cl) <- succ ss;
-  i.reg.loc <- Stack(Local ss);
-  i.reg.spill <- true
+  let ci = active.(cl) in
+  let loc =
+    match
+      List.find_map (fun j ->
+          if Interval.overlap j i then None
+          else begin
+            match j.reg.loc with
+            | Stack(Local _) as loc -> Some loc
+            | _ -> None
+          end
+        ) ci.ci_spilled
+    with
+    | None ->
+        let ss = num_stack_slots.(cl) in
+        num_stack_slots.(cl) <- succ ss;
+        Stack(Local ss)
+    | Some loc -> loc
+  in
+  i.reg.loc <- loc;
+  i.reg.spill <- true;
+  ci.ci_spilled <- i :: ci.ci_spilled
```

Comment:
IIUC, if two "intervals" are mapped to the same stack slot, each of them will appear in ci_spilled; and reusing a stack slot is allowed as soon as one of the intervals using that slot is disjoint with the next interval to be allocated.  I would have expected that to reuse one stack slot, one should ensure that all intervals mapped to it are disjoint from the next interval, no?

Couldn't we reuse the logic in allocate_free_register, and also search for an already allocated stack pseudo-reg in addition to the real registers?

Reply:
Argh, yes, you are right... I'm looking into it.

Reply:
Naive question (I've never seen this code beefore): is this algorithm still "linear"? It looks like N calls to `allocate_slack_slot` (with overlapping intervals) would perform O(N^2) operations.